demos/shadow/shadow_demo_main/shadow_demo_main.c

```diff
@@ -0,0 +1,644 @@
+         * different approach. 
+         */
+
+        /* Then we start to get the version value by JSON keyword "version".
+         * The payload will look similar to this:
+         * {
+         *      "version": 12,
+         *      "timestamp": 1595437367,
+         *      "state": {
+         *          "powerOn": 1
+         *      },
+         *      "metadata": {
+         *          "powerOn": {
+         *          "timestamp": 1595437367
+         *          }
+         *      },
+         *      "clientToken": "388062"
+         *  }
+         */
+        JSONStatus_t result = JSONSuccess;
+
+        result = JSON_Search( ( char * ) pPublishInfo->pPayload,
+                              pPublishInfo->payloadLength,
+                              "version",
+                              sizeof("version") - 1,
+                              '.',
+                              & outValue,
+                              ( size_t * ) & outValueLength );
+        if ( result == JSONSuccess )
+        {
+            LogInfo( ( "version: %.*s\n\n",
+                        outValueLength,
+                        outValue ) );
+
+            /* Convert the code to an unsigned integer value. */
+            version = ( uint32_t ) strtoul( outValue, NULL, 10 );
+
+            LogInfo( ( "version:%d, currentVersion:%d \r\n", version,  currentVersion) );
+
+            if ( version > currentVersion )
+            {
+                /* Set to received version as the current version. */
+                currentVersion = version;
+
+                /* Get powerOn state from json documents. */
+                result = JSON_Search( ( char * ) pPublishInfo->pPayload,
+                                      pPublishInfo->payloadLength,
+                                      "state.powerOn",
+                                      sizeof("state.powerOn") - 1,
+                                      '.',
+                                      & outValue,
+                                      ( size_t * ) & outValueLength );
+
+                if ( result == JSONSuccess )
+                {
+                    /* Convert the code to an unsigned integer value. */
+                    newState = ( uint32_t ) strtoul( outValue, NULL, 10 );
+
+                    LogInfo( ( "The new power on state newState:%d, currentPowerOnState:%d \r\n",
+                               newState,  currentPowerOnState) );
+
+                    if ( newState != currentPowerOnState )
+                    {
+                        currentPowerOnState = newState;
+
+                        /* State change will be handled in main(), where we will publish a "reported"
+                        * state to the device shadow. We do not do it here because we are inside of
+                        * a callback from the MQTT library, so that we don't re-enter
+                        * the MQTT library (even though the library might support re-entrance.) */
```

Comment:
the MQTT library (even though the library might support re-entrance.)  --> the MQTT library.